### PR TITLE
Add a new ACP and group allowing self-service mail updates

### PR DESCRIPTION
**Pull request by euank**
This adds a new "idm_people_self_write_mail_priv" group which follows
the existing canned group+acp format closely.

This also adds a test for the functionality

See the discussion in #648 for a bit more background

- [x] cargo fmt has been run
- [x] cargo clippy has been run
- [x] cargo test has been run and passes
- [ ] ~book chapter included (if relevant)~ | minor doc section added where "people" accounts are mentioned
- [ ] ~design document included (if relevant)~

------

As some meta-commentary, the acp/group code was straightforward enough to make this change feel pretty easy, woo! It's very clearly meant as an intermediate state, but even so it wasn't hard to work with.

The `_UUID` consts, presumably to reserve IDs, did feel a little weird. They weren't actually referenced anywhere, like inline in ACP poicies (rather the UUID itself was copy-pasted), and they don't actually validate uniqueness on their own, other than by having one file to search through.

I'm curious, is the intent just a place to grep for IDs in? Is there a different plan for those consts eventually?

### CODE REVIEW

File: kanidmd/src/lib/constants/acp.rs
Comment (yaleman) [inline kanidmd/src/lib/constants/acp.rs:138]: Should this duplicate `acp_targetscope` be here? 🤔

**Review by yaleman — CHANGES_REQUESTED** [review]

File: kanidmd/src/lib/constants/acp.rs
Comment (Firstyear) [reply kanidmd/src/lib/constants/acp.rs:138]: No, it shouldn't but the second one will over-ride it.

File: kanidmd/src/lib/constants/acp.rs
Comment (Firstyear) [inline kanidmd/src/lib/constants/acp.rs:141]: I think it should be :

```
"{\"and\": [{\"eq\": [\"class\",\"person\"]}, {\"eq\": [\"class\",\"account\"]} , \"self\"]}"
```

Since a person can exist as a phone-book style entry, but can't authenticate, so only accounts (can auth) and people (are human) can update mail.

File: kanidm_book/src/accounts_and_groups.md
Comment (Firstyear) [inline kanidm_book/src/accounts_and_groups.md:146]: Kanidm allow a person to change their legal name, display name, username .... :)

**Review by Firstyear — CHANGES_REQUESTED** [review]

**Comment by Firstyear** [comment]
> As some meta-commentary, the acp/group code was straightforward enough to make this change feel pretty easy, woo! It's very clearly meant as an intermediate state, but even so it wasn't hard to work with.

Yep, it's modeled off of other access control systems :) 

In the future I want to change these around a bit so that these acp rules are "nicer" for the human, but then the actual access is generated out onto the actual entries that recieve because then it's faster to lookup and easier to audit. But right now other things have taken priority :) 

> 
> The `_UUID` consts, presumably to reserve IDs, did feel a little weird. They weren't actually referenced anywhere, like inline in ACP poicies (rather the UUID itself was copy-pasted), and they don't actually validate uniqueness on their own, other than by having one file to search through.

That's pretty much what it is today is a registry of what we allocated. 

> 
> I'm curious, is the intent just a place to grep for IDs in? Is there a different plan for those consts eventually?

Yes, there is a plan for me to fix this up so you can use the group names rather than the consts in there. 


This is pretty much why I haven't exposed a lot of this to the user yet, is I want to rework these internals quite a bit and don't want people to add too many customised access controls today. :)

File: kanidmd/src/lib/constants/acp.rs
Comment (euank) [reply kanidmd/src/lib/constants/acp.rs:138]: It's a copy-paste error, thanks for catching that @yaleman

Fixed, removed the empty one.

File: kanidm_book/src/accounts_and_groups.md
Comment (euank) [reply kanidm_book/src/accounts_and_groups.md]: I was just thinking of person-type attributes, which I thought were just mail and legalname, while display name, username, etc are account attributes (iiuc?)

You're right that the wording's bad. I updated it to not mention legal name specifically, and just call out that mail isn't in the default write set.

File: kanidmd/src/lib/constants/acp.rs
Comment (euank) [reply kanidmd/src/lib/constants/acp.rs]: Sure, updated to that. That makes some sense.

It does seem like it might be inconsistent with the ACP above though, `idm_self_acp_write`. That one has `acp_targetscope: ["self"]`, and allows modifying `legalname`, among other things, with no requirement of `class: "account"` (or not that I see anyway?)

With my current understanding, I also don't see how anything but an account could auth and thus be subject to this ACP.

Is this change to prevent some actual thing now, or does it fit the model better, but in practice not have an observable difference now?
If these phone-book style "people" entries existed (or do they?), would the ACP be able to apply to them in some way?

**Comment by euank** [comment]
Thanks for the answers/clarification, appreciated!

> I want to rework these internals quite a bit and don't want people to add too many customised access controls today. :)

Don't mind me just sneaking this one in... :wink:

------

Anyway, I think I've addressed the review comments, thanks for the review!

**Review by Firstyear — APPROVED** [review]

**Comment by Firstyear** [comment]
No, thank you for contributing!

File: kanidmd/src/lib/constants/acp.rs
Comment (Firstyear) [reply kanidmd/src/lib/constants/acp.rs]: Actually, we should fix that too then 😅

Comment (euank) [reply kanidmd/src/lib/constants/acp.rs]: Sure, done, pushed as an additional commit to this branch.

It would be nice to write a unit test too if it's possible to create a non-account-class "self" that can also call the API.. which I guess is what I'm asking above already. I'm happy to wait for an answer there, or do more digging through code/docs if you've got pointers.

Comment (Firstyear) [reply kanidmd/src/lib/constants/acp.rs]: I don't think you can, not without some of the internal server magic methods. The server internally has the ability to impersonate *any* entry of the system, which allows testing some of these properties, but they aren't exposed (yet) to the API as the security properties and limits on that need clearer definition and design. There are already self write tests, so lets just leave them as is :)

**Comment by Firstyear** [comment]
@yaleman You need to also do an approve (unless I want to be naughty and bypass you :P )

**Comment by Firstyear** [comment]
Don't worry about the test failure, that was my fault, so I'll merge it and fix later :)

**Review by yaleman — APPROVED** [review]

**Comment by yaleman** [comment]
Push da buttonz @Firstyear

**Comment by Firstyear** [comment]
Thank you so much @euank !

**Pull request merged**
